xdsl/dialects/x86/ops.py

```diff
@@ -681,6 +681,144 @@ class RI_MovOp(R_RI_Operation[GeneralRegisterType]):
     name = "x86.ri.mov"
 
 
+class M_MR_Operation(Generic[R1InvT, R2InvT], IRDLOperation, X86Instruction, ABC):
+    """
+    A base class for x86 operations that have one memory reference and one register.
+    """
+
+    r1 = operand_def(R1InvT)
+    r2 = operand_def(R2InvT)
+    offset: AnyIntegerAttr | None = opt_attr_def(AnyIntegerAttr)
+
+    def __init__(
+        self,
+        r1: Operation | SSAValue,
+        r2: Operation | SSAValue,
+        offset: int | AnyIntegerAttr | None,
+        *,
+        comment: str | StringAttr | None = None,
+    ):
+        if isinstance(offset, int):
+            offset = IntegerAttr(offset, 64)
+        if isinstance(comment, str):
+            comment = StringAttr(comment)
+
+        super().__init__(
+            operands=[r1, r2],
+            attributes={
+                "offset": offset,
+                "comment": comment,
+            },
+            result_types=[],
+        )
+
+    def assembly_line_args(self) -> tuple[AssemblyInstructionArg | None, ...]:
+        return self.r1, self.r2
+
+    @classmethod
+    def custom_parse_attributes(cls, parser: Parser) -> dict[str, Attribute]:
+        attributes = dict[str, Attribute]()
+        temp = _parse_optional_immediate_value(
+            parser, IntegerType(64, Signedness.SIGNED)
+        )
+        if temp is not None:
+            attributes["offset"] = temp
+        return attributes
+
+    def custom_print_attributes(self, printer: Printer) -> Set[str]:
+        printer.print(", ")
+        if self.offset is not None:
+            _print_immediate_value(printer, self.offset)
+        return {"offset"}
+
+    def assembly_line(self) -> str | None:
+        instruction_name = self.assembly_instruction_name()
+        destination = _assembly_arg_str(self.r1)
+        source = _assembly_arg_str(self.r2)
+        if self.offset is not None:
+            offset = _assembly_arg_str(self.offset)
+            if self.offset.value.data > 0:
+                return _assembly_line(
+                    instruction_name,
+                    f"[{destination}+{offset}], {source}",
+                    self.comment,
+                )
+            else:
+                return _assembly_line(
+                    instruction_name, f"[{destination}{offset}], {source}", self.comment
+                )
+        else:
+            return _assembly_line(
+                instruction_name, f"[{destination}], {source}", self.comment
+            )
```

Comment:
Could you give a go at doing this via the assembly line args instead? The method you implemented above seems to have different behaviour, and will never be called.

Reply:
hm, not sure how I could represent the memory access format [register + offset] with just assembly_line_args. Also not sure what you mean by saying this method will never be called?

Reply:
`assembly_line_args` is called in the superclass, to construct the assembly line. If `assembly_line` is overridden, then `assembly_line_args` is not called during assembly emission.

You can construct the memory access format into a string in the `assembly_line_args` method, and return that as one of the args, seems like it would be marginally neater.